volumemgr: allow OCI registry fallback datastores

europaul opened the pull request:
A content tree backed by an OCI registry was rejected whenever it referenced more than one datastore. That check assumed an OCI registry can only ever be the single datastore, but the real constraint is narrower: the download path already passes the whole datastore list on as fallbacks, so multiple OCI registries work fine. The only combination that is genuinely unsupported is mixing an OCI registry with non-OCI datastores.

This reworks the check to reject only that mixed case (per @eriknordmark's review suggestion), so a list of OCI registry fallback datastores is accepted as intended. A new `ContentTreeStatus.AllDatastoresAreOCIRegistry()` helper enforces the all-or-none-OCI rule, with unit-test coverage.

For the original (overly conservative) restriction see commit 34ae2287ea6d916e3218a8d84773c69577e0cd3e.

eriknordmark reviewed (commented):
Re-reading the old commit you cited, I think it is overly conservative.
The logic in it assumes that if one datastore is an OCI registry then all of them should be, but it doesn't explicitly check for that but instead checks that there is only one.

I don't think it would be hard to fix that by checking that all or none are OCI. Can you look into this as an alternative approach?

OhmSpectator reviewed (commented):
@europaul what's the status of this PR?

europaul commented:
I'm going to do what @eriknordmark suggested, but I haven't started yet.

OhmSpectator commented:
> I'm going to do what @eriknordmark suggested, but I haven't started yet.

Got it, thanks for the clarification!

rene commented:
@claude

europaul committed:
volumemgr: allow OCI registry fallback datastores

A content tree backed by an OCI registry was rejected whenever it
referenced more than one datastore. That check assumed that an OCI
registry can only ever be the single datastore, but the real
constraint is narrower: the download path already passes the whole
datastore list on as fallbacks, so multiple OCI registries work fine.
The only combination that is genuinely unsupported is mixing an OCI
registry with non-OCI datastores.

Reject only that mixed case instead of any multi-datastore OCI content
tree, so a list of OCI registry fallbacks is accepted as intended.

For the original (overly conservative) restriction see commit
34ae2287ea6d916e3218a8d84773c69577e0cd3e.

Signed-off-by: Paul Gaiduk <paulg@zededa.com>

europaul commented:
I reworked the PR as @eriknordmark suggested. Please give it another look.

rene commented:
@claude

rene commented on pkg/pillar/cmd/volumemgr/updatestatus.go:
```diff
@@ -33,8 +33,11 @@ func doUpdateContentTree(ctx *volumemgrContext, status *types.ContentTreeStatus)
+			// An OCI registry datastore can be combined with fallback
+			// datastores, but only if all of them are OCI registries too.
+			// Mixing OCI and non-OCI datastores is not supported.
+			if !status.AllDatastoresAreOCIRegistry() {
```
@claude , propose the full solution

rene commented:
@claude

eriknordmark reviewed (changes requested):
Makes sense to do the suggested tag resolve fallback as well.
